src/routes/(blog-article)/containerizing-rust-applications-best-practices/+page.md

```diff
@@ -0,0 +1,1156 @@
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+
+## Compile su-exec
+FROM docker.io/library/gcc:bookworm as gcc
+COPY ./contrib/dev-tools/su-exec/ /usr/local/src/su-exec/
+RUN cc -Wall -Werror -g /usr/local/src/su-exec/su-exec.c -o /usr/local/bin/su-exec; chmod +x /usr/local/bin/su-exec
+
+## Application
+FROM rust:1.73 as builder
+WORKDIR /app
+RUN cargo init
+RUN cargo build --release
+CMD ["./target/release/app"]
+
+## Runtime
+FROM gcr.io/distroless/cc-debian12:debug as runtime
+RUN ["/busybox/cp", "-sp", "/busybox/sh","/busybox/cat","/busybox/ls","/busybox/env", "/bin/"]
+
+COPY --from=builder /app/target/release/app /usr/local/bin/app
+COPY --from=gcc --chmod=0555 /usr/local/bin/su-exec /bin/su-exec
+COPY --chmod=0555 ./share/container/entry_script_sh /usr/local/bin/entry.sh
+
+ENTRYPOINT ["/usr/local/bin/entry.sh"]
+CMD ["/usr/local/bin/app"]
+
+```
+
+</CodeBlock>
+
+This is the approach we use in Torrust. You run the docker as `root` but we always
+use an entrypoint. That entrypoint creates a new user with an ID provided as an
+environment variable.
+
+The `entry.sh` script is always called when you run the container because it's
+defined as an `ENTRYPOINT`. This "middleware" script creates the user if it does
+not exit and then runs the application using the [su-exec](https://github.com/ncopa/su-exec) program.
+
+As you can read on the su-exec documentation:
+
+<Callout type="info">
+
+"su-exec" is a simple tool that will simply execute a program with different privileges.
+The program will be executed directly and not run as a child, like su and sudo does,
+which avoids TTY and signal issues (see below).
+
+Notice that su-exec depends on being run by the root user, non-root users do not
+have permission to change uid/gid.
+
+</Callout>
+
+The advantage of this approach is that you don't have to worry about the user ID
+when you build the docker image. You can run the container as root and then use
+the entrypoint to create the user and run the application with that user.
+
+The entrypoint also ensures that the application is executed with the correct
+permissions. You cannot run the application as root unless you explicitly set
+the environment variable `USER_ID` to `0`.
+
+If you want to contribute with Torrust we think that we could simplify the Containerfile
+if the "su-exec" command were available in Rust. Because we could reuse one of the
+Rust docker images we are using for other docker stages.
+
+## Decomposing the Torrust Tracker Containerfile
+
+Finally we can explain line by line what the [Torrust Tracker Containerfile](https://github.com/torrust/torrust-tracker/blob/develop/Containerfile) does.
+
+Rust apps can be build in `debug` or `release` mode. The `Containerfile` builds both
+modes and then it runs the tests for both modes. We have removed the `debug` mode
+to keep the example short. But it's almost the same code. FOr the `release` mode
+the flag `--release` is added to some commands.
+
+This is the full example:
```

Comment:
I sugest that we move this example to replace our Containerfile at the start of the blog post. We can link to the latest version of our actual Containerfile in the repository.

Reply:
Done.